pootle/apps/pootle_fs/finder.py

```diff
@@ -0,0 +1,182 @@
+        if not matched.get("filename"):
+            matched["filename"] = os.path.splitext(
+                os.path.basename(filename))[0]
+        return file_path, matched
+
+    def walk(self):
+        # print("Walking the FS: %s" % self.file_root)
+        for root, dirs, files in scandir.walk(self.file_root):
+            for filename in files:
+                yield os.path.join(root, filename)
+
+    def find(self):
+        for filepath in self.walk():
+            match = self.match_filepath(filepath)
+            if match:
+                yield match
+
+    def match(self, file_path):
+        return self.regex.match(file_path)
+
+    @lru_cache(maxsize=None)
+    def reverse_match(self, language_code, filename=None, ext=None, dir_path=None):
+        if not filename:
+            filename = language_code
+        if ext is None:
+            ext = self.ext[0]
+        if ext not in self.ext:
+            raise ValueError("ext must be in the list of possible extensions")
+        ext = ext.strip(".")
+        path = (
+            "%s.%s"
+            % (os.path.splitext(self.translation_path)[0], ext))
+        if dir_path and "<dir_path>" not in path:
+            return
+        path = (path.replace("<language_code>", language_code)
+                    .replace("<filename>", filename))
+        if "<dir_path>" in path:
+            if dir_path and dir_path.strip("/"):
+                path = path.replace(
+                    "<dir_path>", "/%s/" % dir_path.strip("/"))
+            else:
+                path = path.replace("<dir_path>", "")
+        local_path = path.replace(self.file_root, "")
+        if "//" in local_path:
+            path = os.path.join(
+                self.file_root,
+                local_path.replace("//", "/").lstrip("/"))
+        return path
+
+    def _ext_re(self):
+        return (
+            r".(?P<ext>(%s))"
+            % "|".join(
+                ("%s$" % x)
+                for x in set(self.ext)))
+
+    def _parse_path_regex(self):
+        path = self.translation_path
+        for k, v in self.path_mapping:
+            path = path.replace(k, v)
+        return "%s%s$" % (
+            os.path.splitext(path)[0],
+            self._ext_re())
+
+    def _parse_path(self):
+        regex = self._parse_path_regex()
+        if not self.path_filters:
+            return regex
+        regex = r"(?=%s)" % regex
+        filter_regex = "".join(
+            [("(?=^%s)" % fnmatch.translate(path))
+             for path in self.path_filters])
+        return "%s%s" % (regex, filter_regex)
+
+
+class TranslationPathValidator(object):
+
+    validators = (
+        "no_dots", "lang_code", "ext", "match_tags", "path")
+
+    def __init__(self, path):
+        self.path = path
+
+    def validate_no_dots(self):
+        links_to_parent = (
+            self.path == ".."
+            or self.path.startswith("../")
+            or "/../" in self.path
+            or self.path.endswith("/.."))
```

Comment:
`foo..bar` is a valid filepath. I suppose (and this relates to valid chars issue too) it deps on how much we want to support of any potential filesystems